## Part 1: Explore the Data

Import the data and use Pandas to learn more about the dataset.

In [1]:
import pandas as pd


clients_df = pd.read_csv('Resources/client_dataset.csv')
clients_df.head()

first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year         item_id     category         subcategory  \
0          17        2023  EUD29711-63-6U        decor            wall art   
1          20        2023  XDA18116-89-4A  consumables                pens   
2           4        2023  ABE59463-05-7E     software  project management   
3          17        2023  ZMM00836-65-0C  consumables                pens   
4          21        2023  BZX55559-12-3X  consumables                misc   

   unit_price  unit_cost  unit_weight  qty  line_number  
0     1096.80     762.71         7.50  105            1  
1       24.95      15.09         1.49   21            0  
2       13.52       7.86         1.68   39            6  
3       36.42      24.85         1.23   29            3  
4      195.10     108.17        46.43   20            1

In [2]:
# View the column names in the data
clients_df.columns.to_list()

['first',
 'last',
 'job',
 'phone',
 'email',
 'client_id',
 'order_id',
 'order_date',
 'order_week',
 'order_year',
 'item_id',
 'category',
 'subcategory',
 'unit_price',
 'unit_cost',
 'unit_weight',
 'qty',
 'line_number']

In [3]:
# Use the describe function to gather some basic statistics
clients_df.describe()
print(clients_df[['client_id', 'order_id', 'order_week', 'order_year', 'unit_price', 'unit_cost', 'unit_weight', 'qty', 'line_number']].describe())


          client_id      order_id    order_week    order_year    unit_price  \
count  54639.000000  5.463900e+04  54639.000000  54639.000000  54639.000000   
mean   54837.869416  5.470190e+06     11.359139   2022.993064    136.267207   
std    25487.438231  2.599807e+06      7.023499      0.082997    183.873135   
min    10033.000000  1.000886e+06      1.000000   2022.000000      0.010000   
25%    33593.000000  3.196372e+06      6.000000   2023.000000     20.800000   
50%    53305.000000  5.496966e+06     11.000000   2023.000000     68.310000   
75%    78498.000000  7.733869e+06     17.000000   2023.000000    173.160000   
max    99984.000000  9.998480e+06     52.000000   2023.000000   1396.230000   

          unit_cost   unit_weight           qty   line_number  
count  54639.000000  54639.000000  5.463900e+04  54639.000000  
mean      99.446073      5.004116  5.702646e+02      2.979667  
std      133.164267      5.326599  1.879552e+04      2.436320  
min        0.010000      0.00000

In [4]:
clients_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 54639 entries, 0 to 54638
Data columns (total 18 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   first        54639 non-null  object 
 1   last         54639 non-null  object 
 2   job          54639 non-null  object 
 3   phone        54639 non-null  object 
 4   email        54639 non-null  object 
 5   client_id    54639 non-null  int64  
 6   order_id     54639 non-null  int64  
 7   order_date   54639 non-null  object 
 8   order_week   54639 non-null  int64  
 9   order_year   54639 non-null  int64  
 10  item_id      54639 non-null  object 
 11  category     54639 non-null  object 
 12  subcategory  54639 non-null  object 
 13  unit_price   54639 non-null  float64
 14  unit_cost    54639 non-null  float64
 15  unit_weight  54639 non-null  float64
 16  qty          54639 non-null  int64  
 17  line_number  54639 non-null  int64  
dtypes: float64(3), int64(6), object(9)
memory usag

In [5]:
# Use this space to do any additional research
# and familiarize yourself with the data.

columns_df = pd.DataFrame(clients_df.columns, columns=['Column Name']) 
columns_df['Count'] = pd.DataFrame(clients_df.count(), columns=['Count']).values
columns_df['Unique'] = pd.DataFrame(clients_df.nunique(), columns=['Unique']).values
columns_df





Column Name  Count  Unique
0         first  54639     360
1          last  54639     477
2           job  54639     498
3         phone  54639     999
4         email  54639     999
5     client_id  54639     999
6      order_id  54639   10000
7    order_date  54639     150
8    order_week  54639      23
9    order_year  54639       2
10      item_id  54639     500
11     category  54639       5
12  subcategory  54639      18
13   unit_price  54639     490
14    unit_cost  54639     487
15  unit_weight  54639     393
16          qty  54639    2911
17  line_number  54639      10

In [6]:
# What three item categories had the most entries?
category_counts_df = clients_df.sort_values('client_id', ascending=False)
category_counts_df = category_counts_df['category'].value_counts()
category_counts_df.head(3)

category
consumables    23538
furniture      11915
software        8400
Name: count, dtype: int64

In [7]:
# For the category with the most entries, which subcategory had the most entries?
subcategory_counts_df = clients_df.sort_values('subcategory', ascending=False)
subcategory_counts_df = subcategory_counts_df['subcategory'].value_counts()
subcategory_counts_df.head(1)  

subcategory
bathroom supplies    6424
Name: count, dtype: int64

In [8]:
# Which five clients had the most entries in the data?
top_clients = clients_df.groupby(by='client_id').count().sort_values(by='category', ascending=False)
count = top_clients['category'].head(5)
count

client_id
33615    220
66037    211
46820    209
38378    207
24741    207
Name: category, dtype: int64

In [9]:
# Store the client ids of those top 5 clients in a list.
top_clients_list = count.index.to_list()
top_clients_list



[33615, 66037, 46820, 38378, 24741]

In [10]:
# How many total units (the qty column) did the client with the most entries order order?
top_client_units = clients_df[clients_df['client_id'] == top_clients_list[0]]
total_units = top_client_units['qty'].sum()
total_units


64313

## Part 2: Transform the Data
Do we know that this client spent the more money than client 66037? If not, how would we find out? Transform the data using the steps below to prepare it for analysis.

In [11]:
# Create a column that calculates the subtotal for each line using the unit_price and the qty
def calculate_subtotal(row):
    return row['unit_price'] * row['qty']

clients_df['subtotal'] = clients_df.apply(calculate_subtotal, axis=1)
print(clients_df[['client_id', 'unit_price', 'qty', 'subtotal']].head())







   client_id  unit_price  qty   subtotal
0      58515     1096.80  105  115164.00
1      37609       24.95   21     523.95
2      57113       13.52   39     527.28
3      46554       36.42   29    1056.18
4      92089      195.10   20    3902.00


In [12]:
# Create a column for shipping price.
# Assume a shipping price of $7 per pound for orders over 50 pounds and $10 per pound for items 50 pounds or under.
def calculate_shipping_price(row):
    if row['unit_weight'] * row['qty'] > 50:
        return row['unit_weight'] * row['qty'] * 7
    else:
        return row['unit_weight'] * row['qty'] * 10
    
clients_df['shipping_price'] = clients_df.apply(calculate_shipping_price, axis=1)
clients_df[['client_id', 'unit_price', 'unit_weight', 'qty', 'shipping_price']].head()


client_id  unit_price  unit_weight  qty  shipping_price
0      58515     1096.80         7.50  105         5512.50
1      37609       24.95         1.49   21          312.90
2      57113       13.52         1.68   39          458.64
3      46554       36.42         1.23   29          356.70
4      92089      195.10        46.43   20         6500.20

In [13]:
# Create a column for the total price using the subtotal and the shipping price along with a sales tax of 9.25%
def calculate_total_price(row):
    return (row['subtotal'] + row['shipping_price']) * 1.0925

clients_df['total_price'] = clients_df.apply(calculate_total_price, axis=1)
clients_df[['client_id', 'unit_price', 'qty', 'subtotal', 'shipping_price', 'total_price']].head()

client_id  unit_price  qty   subtotal  shipping_price    total_price
0      58515     1096.80  105  115164.00         5512.50  131839.076250
1      37609       24.95   21     523.95          312.90     914.258625
2      57113       13.52   39     527.28          458.64    1077.117600
3      46554       36.42   29    1056.18          356.70    1543.571400
4      92089      195.10   20    3902.00         6500.20   11364.403500

In [14]:
# Create a column for the cost of each line using unit cost, qty, and
# shipping price (assume the shipping cost is exactly what is charged to the client).
def calculate_line_cost(row):
    return (row['unit_cost'] * row['qty']) + row['shipping_price']
                
clients_df['line_cost'] = clients_df.apply(calculate_line_cost, axis=1)
clients_df[['client_id', 'unit_price', 'unit_weight', 'qty', 'shipping_price', 'line_cost']].head()


client_id  unit_price  unit_weight  qty  shipping_price  line_cost
0      58515     1096.80         7.50  105         5512.50   85597.05
1      37609       24.95         1.49   21          312.90     629.79
2      57113       13.52         1.68   39          458.64     765.18
3      46554       36.42         1.23   29          356.70    1077.35
4      92089      195.10        46.43   20         6500.20    8663.60

In [15]:
# Create a column for the profit of each line using line cost and line price
def calculate_line_profit(row):
    return row['total_price'] - row['line_cost']

clients_df['line_profit'] = clients_df.apply(calculate_line_profit, axis=1)
clients_df[['client_id', 'unit_price', 'unit_weight', 'qty', 'shipping_price', 'line_cost', 'line_profit']].head()


client_id  unit_price  unit_weight  qty  shipping_price  line_cost  \
0      58515     1096.80         7.50  105         5512.50   85597.05   
1      37609       24.95         1.49   21          312.90     629.79   
2      57113       13.52         1.68   39          458.64     765.18   
3      46554       36.42         1.23   29          356.70    1077.35   
4      92089      195.10        46.43   20         6500.20    8663.60   

    line_profit  
0  46242.026250  
1    284.468625  
2    311.937600  
3    466.221400  
4   2700.803500

In [16]:
clients_df

first       last                      job               phone  \
0         Donald    Harding             Immunologist  793-904-7725x39308   
1        Tiffany      Myers          Music therapist    201.442.4543x942   
2        Shannon     Watson             Immunologist   687.737.9424x8503   
3         Nathan      Baker    Accounting technician    827-788-8123x012   
4      Christina   Schwartz             Chiropractor        265-829-3643   
...          ...        ...                      ...                 ...   
54634     Lauren      Reese           Radio producer  977.877.5272x11382   
54635    Derrick      Moore                 Musician        358.661.5483   
54636     Monica  Gutierrez         Graphic designer    294.805.9100x339   
54637      Wanda    Solomon             Toxicologist       (311)767-4924   
54638     Dwayne       Bell  Chief Financial Officer   606-619-7667x9508   

                                     email  client_id  order_id  order_date  \
0         harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1        myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                    swatson8146@payne.net      57113   1902144  2023-01-29   
3                   bakernathan@benson.com      46554   9031802  2023-04-25   
4      christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   
...                                    ...        ...       ...         ...   
54634   late_reese_4081@montoya-chavez.org      29043   9021716  2023-04-26   
54635         derrick.moore.2602@pope.info      41908   6290153  2023-02-08   
54636            gutierrezm3195@morris.org      35176   8692622  2023-03-05   
54637            solomonwanda5962@ross.org      24485   7592730  2023-01-18   
54638                     belld@galvan.com      24007   7489403  2023-01-04   

       order_week  order_year  ... unit_price unit_cost unit_weight   qty  \
0              17        2023  ...    1096.80    762.71        7.50   105   
1              20        2023  ...      24.95     15.09        1.49    21   
2               4        2023  ...      13.52      7.86        1.68    39   
3              17        2023  ...      36.42     24.85        1.23    29   
4              21        2023  ...     195.10    108.17       46.43    20   
...           ...         ...  ...        ...       ...         ...   ...   
54634          17        2023  ...      83.13     51.60        2.25    33   
54635           6        2023  ...     206.59    175.46       11.70    47   
54636           9        2023  ...      65.66     57.31        4.16   475   
54637           3        2023  ...       1.48      1.22       18.04   112   
54638           1        2023  ...       3.01      1.61        2.07  1031   

       line_number   subtotal  shipping_price    total_price  line_cost  \
0                1  115164.00         5512.50  131839.076250   85597.05   
1                0     523.95          312.90     914.258625     629.79   
2                6     527.28          458.64    1077.117600     765.18   
3                3    1056.18          356.70    1543.571400    1077.35   
4                1    3902.00         6500.20   11364.403500    8663.60   
...            ...        ...             ...            ...        ...   
54634            8    2743.29          519.75    3564.871200    2222.55   
54635            0    9709.73         3849.30   14813.240275   12095.92   
54636            6   31188.50        13832.00   49184.896250   41054.25   
54637            0     165.76        14143.36   15632.713600   14280.00   
54638            5    3103.31        14939.19   19711.431250   16599.10   

        line_profit  
0      46242.026250  
1        284.468625  
2        311.937600  
3        466.221400  
4       2700.803500  
...             ...  
54634   1342.321200  
54635   2717.320275  
54636   8130.646250  
54637   1352.713600  
54638   3112.331250  

[54639 rows x 23 columns]

## Part 3: Confirm your work
You have email receipts showing that the total prices for 3 orders. Confirm that your calculations match the receipts. Remember, each order has multiple lines.

Order ID 2742071 had a total price of \$152,811.89

Order ID 2173913 had a total price of \$162,388.71

Order ID 6128929 had a total price of \$923,441.25


In [17]:
# Check your work using the totals above
order_list = [2742071, 2173913, 6128929]
for order in order_list:
    total_value = clients_df.loc[clients_df['order_id'] == order, 'total_price'].sum()
    print (f'Total value for order {order} is: {total_value:.2f}')



Total value for order 2742071 is: 152811.90
Total value for order 2173913 is: 162388.72
Total value for order 6128929 is: 923441.24


## Part 4: Summarize and Analyze
Use the new columns with confirmed values to find the following information.

In [18]:
# How much did each of the top 5 clients by quantity spend? Check your work from Part 1 for client ids.
for client in top_clients_list:
    total_value = clients_df.loc[clients_df['client_id'] == client, 'total_price'].sum()
    print (f'Total value for client {client} is: {total_value:.2f}')


Total value for client 33615 is: 8377308.57
Total value for client 66037 is: 10259514.80
Total value for client 46820 is: 9743794.32
Total value for client 38378 is: 12906550.88
Total value for client 24741 is: 82268891.98


In [19]:
# Create a summary DataFrame showing the totals for the for the top 5 clients with the following information:
# total units purchased, total shipping price, total revenue, and total profit. 
top_five_clients = {"client_id": top_clients_list}
columns_list = ['qty', 'shipping_price', 'total_price', 'line_cost', 'line_profit']
for column in columns_list:
       top_five_clients[column]= []    
for client in top_clients_list:
        for column in columns_list:
                total_value = clients_df.loc[clients_df['client_id'] == client, column].sum()
                top_five_clients[column].append(total_value)
                
summary_df = pd.DataFrame(top_five_clients)
summary_df



client_id     qty  shipping_price   total_price    line_cost   line_profit
0      33615   64313      1828984.89  8.377309e+06   6175313.91  2.201995e+06
1      66037   43018      1395151.85  1.025951e+07   7004482.98  3.255032e+06
2      46820   75768      1601448.84  9.743794e+06   7007191.64  2.736603e+06
3      38378   73667      3429455.40  1.290655e+07   9634720.98  3.271830e+06
4      24741  239862      5126448.37  8.226889e+07  45688899.71  3.657999e+07

In [20]:
# Format the data and rename the columns to names suitable for presentation.

# Define the money columns. 

# Define a function that converts a dollar amount to millions.

# Apply the currency_format_millions function to only the money columns. 

# Rename the columns to reflect the change in the money format. 

columns = ['shipping_price', 'total_price', 'line_cost', 'line_profit']
for column in columns:
    summary_df[column] = summary_df[column] / 1000000

renamed_df = summary_df.rename(columns = {
    "shipping_price": "Shipping Price (millions)",
    "total_price": "Total Revenue (millions)",
    "line_cost": "Total Cost (millions)", 
    "line_profit": "Total Profit (millions)"
})      

renamed_df.head()



client_id     qty  Shipping Price (millions)  Total Revenue (millions)  \
0      33615   64313                   1.828985                  8.377309   
1      66037   43018                   1.395152                 10.259515   
2      46820   75768                   1.601449                  9.743794   
3      38378   73667                   3.429455                 12.906551   
4      24741  239862                   5.126448                 82.268892   

   Total Cost (millions)  Total Profit (millions)  
0               6.175314                 2.201995  
1               7.004483                 3.255032  
2               7.007192                 2.736603  
3               9.634721                 3.271830  
4              45.688900                36.579992

In [21]:
# Sort the updated data by "Total Profit (millions)" form highest to lowest and assign the sort to a new DatFrame.
renamed_df = renamed_df.sort_values(by="Total Profit (millions)", ascending=False)
renamed_df.head()


client_id     qty  Shipping Price (millions)  Total Revenue (millions)  \
4      24741  239862                   5.126448                 82.268892   
3      38378   73667                   3.429455                 12.906551   
1      66037   43018                   1.395152                 10.259515   
2      46820   75768                   1.601449                  9.743794   
0      33615   64313                   1.828985                  8.377309   

   Total Cost (millions)  Total Profit (millions)  
4              45.688900                36.579992  
3               9.634721                 3.271830  
1               7.004483                 3.255032  
2               7.007192                 2.736603  
0               6.175314                 2.201995

In [22]:
#**Summary**:

#Analysis of the wholesale data show that total profit is not linearly proportional to units sold.
# as evident from the top 5 clients.  For example, client 2742071 has the highest total units sold, 
# with a 79% profit margin.  #Whereas, client 3361512 has lower number of units sold, but at 
# only 35% profit margin.  #The data should be reviewed by Sales for further analysis and corrective actions.